### PR TITLE
Give EC2 instances a user-data script to run

### PR DESCRIPTION
**Pull request by wckzhang**
Signed-off-by: William Zhang <wilzhang@amazon.com>

*Issue #, if available:*
N/A

*Description of changes:*
Passes in the user-data script to the ec2 instances

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: common.sh [hunk 1]
```diff
@@ -216,6 +216,7 @@ create_instance()
                     --network-interface ${network_interface}",\"SubnetId\":\"${subnet}\"}]" \
                     --count ${NODES}:${NODES} \
                     --query "Instances[*].InstanceId" \
+                    --user-data file://${JENKINS_HOME}/user_data_script.sh \
```
Comment (zhngaj) [inline common.sh:219]: Just want to double check is it guaranteed that JENKINS_HOME is defined?

Comment (wckzhang) [reply common.sh]: I updated it, I checked and there are multiple consumers of create instances, so I changed it to an environment variable argument.

File: multi-node.sh [hunk 2]
```diff
@@ -11,6 +11,7 @@ libfabric_job_type=${libfabric_job_type:-"master"}
 # Current LibfabricCI IAM permissions do not allow placement group creation,
 # enable this after it is fixed.
 # export ENABLE_PLACEMENT_GROUP=1
+export USER_DATA_FILE=${JENKINS_HOME}/user_data_script.sh
```
Comment (shijin-aws) [inline multi-node.sh:14]: These scripts are also used by installer pipeline's test stage, where `$JENKINS_HOME` is not available.

Comment (shijin-aws) [reply multi-node.sh:14]: You can set this as a default value but can be overridden outside the scripts.

File: single-node.sh [hunk 3]
```diff
@@ -9,6 +9,7 @@ ami=($slave_value)
 NODES=1
 # Placement group is not needed for single-node tests.
 export ENABLE_PLACEMENT_GROUP=0
+export USER_DATA_FILE=${JENKINS_HOME}/user_data_script.sh
```
Comment (shijin-aws) [inline single-node.sh:12]: Same comment as in `multi-node.sh`

File: nccl/common/nccl-common.sh [hunk 4]
```diff
@@ -169,6 +172,7 @@ create_instance() {
                     --network-interface ${network_interface}",\"SubnetId\":\"${subnet}\"}]" \
                     --count $2 \
                     --query "Instances[*].InstanceId" \
+                    --user-data file://${JENKINS_HOME}/user_data_script.sh \
```
[the remark this answers is not in the document]
Comment (wckzhang) [reply nccl/common/nccl-common.sh:175]: Need to remove thanks

File: multi-node.sh (hunk 2 above)
Comment (wckzhang) [reply multi-node.sh:14]: I'll ping you directly

File: multi-node.sh [hunk 5]
```diff
@@ -11,6 +11,7 @@ libfabric_job_type=${libfabric_job_type:-"master"}
 # Current LibfabricCI IAM permissions do not allow placement group creation,
 # enable this after it is fixed.
 # export ENABLE_PLACEMENT_GROUP=1
+export USER_DATA_FILE=${USER_DATA_FILE:-${JENKINS_HOME}/user_data_script.sh}
```
Comment (zhngaj) [inline multi-node.sh:14]: Both Shi and I had a concern about variable ${JENKINS_HOME}, just to make sure that you two have sync up, and already had an agreement here.

Comment (shijin-aws) [reply multi-node.sh:14]: I have talked with @wckzhang offline about this and this change looks good to me.

Comment (wckzhang) [reply multi-node.sh:14]: Yes, so there are a couple checks now. If USER_DATA_FILE exists, use that, if it doesn't use JENKINS_HOME/user_data_script.sh If this file doesn't exist, the run-instances command won't use this argument and there's no harm.

**Review by shijin-aws — APPROVED** [review]

**Pull request merged**